docs/admin/configuration.md
## Configuration

There are two ways of configuring Perseo CEP:

-   The default Basic Configuration is read from the `config.js` file in the root of the project folder.
-   Some pieces of configuration can be overriden using environment variables, as it is explained in the following
    section.

### Environment Variables Configuration

The following table shows the environment variables available for Perseo configuration:

| Environment variable                 | Description                                                                                                                                                                                                                                                                   |
| :----------------------------------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| PERSEO_ENDPOINT_HOST                 | Host where the CEP will listen.                                                                                                                                                                                                                                               |
| PERSEO_ENDPOINT_PORT                 | Port where the CEP will listen.                                                                                                                                                                                                                                               |
| PERSEO_MONGO_URI                     | URI for Mongo DB. This option overrides the Mongo options below, which have been deprecated. Mongodb URI format is available in `http://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html`                                                               |
| PERSEO_IS_MASTER                     | true if this one is the master or false it is the slave                                                                                                                                                                                                                       |
| PERSEO_SLAVE_DELAY                   | Slave's delay to try to execute an action (milliseconds).                                                                                                                                                                                                                     |
| PERSEO_CORE_URL                      | Full URL where Perseo Core is listening (e.g: `http://63.34.124.1:8080`).                                                                                                                                                                                                     |
| PERSEO_NEXT_URL                      | Full URL where Perseo Core replicated node is listening. Same format as above.                                                                                                                                                                                                |
| PERSEO_ORION_URL                     | Full URL of the Orion Context Broker (e.g: `http://64.124.28.15:1026`).                                                                                                                                                                                                       |
| PERSEO_PEP_URL                       | Full URL of the Pep (Context Broker) (e.g: `http://64.124.28.15:10026`). This URL will be used instead of PERSEO_ORION_URL when updateAction rule has set a non empty value for service and subservice.                                                                       |
| PERSEO_LOG_LEVEL                     | Log level.                                                                                                                                                                                                                                                                    |
| PERSEO_SMTP_HOST                     | Host of the SMTP server                                                                                                                                                                                                                                                       |
| PERSEO_SMTP_PORT                     | Port of the SMTP server                                                                                                                                                                                                                                                       |
| PERSEO_SMTP_SECURE                   | `true` if SSL should be used with the SMTP server                                                                                                                                                                                                                             |
| PERSEO_SMTP_AUTH_USER                | Authentication data, the username                                                                                                                                                                                                                                             |
| PERSEO_SMTP_AUTH_PASS                | Authentication data, the password for the user                                                                                                                                                                                                                                |
| PERSEO_SMTP_TLS_REJECTUNAUTHORIZED   | Reject if unauthorized security is found (i.e. selfsigned certificates). Default is false.                                                                                                                                                                                    |
| PERSEO_SMTP_TLS_CA                   | CA to override the trusted CA certificates. Default is empty. Check more detail in https://nodejs.org/api/tls.html#class-tlstlssocket                                                                                                                                         |
| PERSEO_SMTP_TLS_KEY                  | Private keys in PEM format. Default is empty. Check more detail in https://nodejs.org/api/tls.html#class-tlstlssocket                                                                                                                                                         |
| PERSEO_SMTP_TLS_CERT                 | Cert chains in PEM format. Default is empty. Check more detail in https://nodejs.org/api/tls.html#class-tlstlssocket                                                                                                                                                          |
| PERSEO_SMTP_TLS_MIN_VERSION          | Optionally set the minimum TLS version to allow. One of 'TLSv1.3', 'TLSv1.2', 'TLSv1.1', or 'TLSv1'. Check more detail in https://nodejs.org/api/tls.html#class-tlstlssocket                                                                                                  |
| PERSEO_SMTP_TLS_SERVER_NAME          | Server name for the SNI (Server Name Indication) TLS extension. Check more detail in https://nodejs.org/api/tls.html#class-tlstlssocket                                                                                                                                       |
| PERSEO_SMS_URL                       | URL for sending SMSs (SMPP Adapter)                                                                                                                                                                                                                                           |
| PERSEO_SMS_API_KEY                   | API KEY for sending SMSs, if necessary. Only for the SMPP Adapter simulator                                                                                                                                                                                                   |
| PERSEO_SMS_API_SECRET                | API SECRET for sending SMSs, if necessary. Only for the SMPP Adapter simulator                                                                                                                                                                                                |
| PERSEO_SMS_FROM                      | Field `from` for the outgoing SMSs. Required by the SMPP Adapter                                                                                                                                                                                                              |
| PERSEO_SMPP_HOST                     | Host of the SMPP server                                                                                                                                                                                                                                                       |
| PERSEO_SMPP_PORT                     | Port of the SMPP server                                                                                                                                                                                                                                                       |
| PERSEO_SMPP_SYSTEMID                 | SystemID for the user of the SMPP server                                                                                                                                                                                                                                      |
| PERSEO_SMPP_PASSWORD                 | Password for the user of the SMPP server                                                                                                                                                                                                                                      |
| PERSEO_SMPP_FROM                     | Number from SMS are sending by SMPP server.                                                                                                                                                                                                                                   |
| PERSEO_SMPP_ENABLED                  | SMPP is default method for SMS instead of use SMS gateway                                                                                                                                                                                                                     |
| PERSEO_SMPP_DEFAULT_COUNTRY_CODE     | SMPP default country code (e.g., '34' for Spain)                                                                                                                                                                                                                              |
| PERSEO_NOTICES_PATH                  | Path for incoming notices, default value '/notices'                                                                                                                                                                                                                           |
| PERSEO_RULES_PATH                    | Path for incoming rules, default value '/rules'                                                                                                                                                                                                                               |
| PERSEO_CAST_TYPE                     | Maximum number of rule executions triggered by the same notification (with the same correlator) until refuse execute a rule. Default is 20. See [loop detection tips](https://github.com/telefonicaid/perseo-fe/blob/master/docs/architecture/architecture.md#loop-detection) |
| PERSEO_MAX_RULES_BY_CORR             | If true, enable attribute value casting based in NGSI-v2 attribute types if true. If false (default), the JSON native type for the attribute value is used.                                                                                                                   |
| PERSEO_CHECK_NON_SIGNAL_BY_API       | If false, the entities will be directly find using MongoDB, if true they will be fing using NGSIv2 API(\*)                                                                                                                                                                    |
| PERSEO_NON_SIGNAL_MAX_TIME_DETECTION | Max time detection for check nonSignal rules by mongo in seconds                                                                                                                                                                                                              |

(\*)Although the idea is that now Perseo works always using NGSIv2 API to check times for non signal rules, this setting
has been introduced to ease a potential rollback if some problem is found with this new approach. Thus, I has to be
considered a kind of temporal debugging settings, to be removed when we get confident with the new functionality, thus
you should use it with care.

### Basic Configuration

In order to have perseo running, there are several basic pieces of information to fill:

-   `config.logLevel`: level for log messages (`FATAL`, `ERROR`, `WARN`, `INFO` or `DEBUG`)
-   `config.perseoCore.rulesURL`: URL for management of EPL rules at core.
-   `config.perseoCore.noticesURL`: URL for processing events at core rule engine.
-   `config.perseoCore.interval`: interval for refreshing rules at core rule engine (milliseconds).
-   `config.smtp.port`: port for sending email.
-   `config.smtp.host`: host for sending email.
-   `config.smtp.secure`: defines if the connection should use SSL (if true) or not (if false).
-   `config.smtp.auth.user`: authentication data, the username.
-   `config.smtp.auth.pass`: authentication data, the password for the user.
-   `config.smtp.tls.rejectUnauthorized`: reject if unauthorized security is found (i.e. selfsigned certificates).
    Default is false.
-   `config.sms.URL`: URL for sending SMSs.
-   `config.sms.from`: Field `from` for the outgoing SMSs. Required by the SMPP Adapter.
-   `config.sms.API_KEY`: API KEY for sending SMSs, if necessary. Only for the SMPP Adapter simulator.
-   `config.sms.API_SECRET`: API SECRET for sending SMSs, if necessary. Only for the SMPP Adapter simulator.
-   `config.smpp.host`: Host of the SMPP server.
-   `config.smpp.port`: Port of the SMPP server.
-   `config.smpp.systemid`: SystemID for the user of the SMPP server
-   `config.smpp.password`: Password for the user of the SMPP server
-   `config.smpp.from`: Number from SMS are sending by SMPP server
-   `config.smpp.enabled`: SMPP is default method for SMS instead of use SMS gateway.
-   `config.smpp.defaultCountryCode`: SMPP default country code (e.g., '34' for Spain).
-   `config.orion.URL`: Context Broker base URL, e.g. `https://orion.example.com:1026`
-   `config.mongo.URL`: URL for connecting mongoDB.
-   `config.executionsTTL`: Time-To-Live for documents of action executions (seconds).
-   `config.checkDB.delay`: Number of milliseconds to check DB connection (see
    [database aspects](admin.md#database-aspects) documentation for mode detail).
-   `config.checkDB.reconnectTries`: Number of attempts to reconnect (see [database aspects](admin.md#database-aspects)
    documentation for mode detail).
-   `config.checkDB.reconnectInterval`: Number of milliseconds to wait between attempts to reconnect (see
    [database aspects](admin.md#database-aspects) documentation for mode detail).
-   `config.checkDB.bufferMaxEntries`: Number of operations buffered up before giving up on getting a working connection
    (see [database aspects](admin.md#database-aspects) documentation for mode detail).
-   `config.maxRulesByCorr`: Maximum number of rule executions triggered by the same notification (with the same
    correlator) until refuse execute a rule. Defualt is 20
-   `config.nonSignalByAPI`: Flag value, determine the use of the NGSIv2 API or MongoDB. By default is false.
-   `config.nonSignalMaxTimeDetection`: Max time detection for check nonSignal rules by mongo in seconds. Default is 30
    days.

Options for HA:

-   `config.isMaster`: `true` if this one is the master or `false` it is the slave.
-   `config.slaveDelay`: Slave's delay to try to execute an action (milliseconds).
-   `config.nextCore.rulesURL`: URL for management of EPL rules at _replicated_ core. If set, the rules will be
    propagated to that one also.
-   `config.nextCore.noticesURL`: URL for processing events at _replicated_ core rule engine. If set, the events will be
    propagated to that one also.

Options for Authentication through PEP (for update action) in the configTrust.js file:

-   `configTrust.trust[].host`: host (keyStone) to exchange tokens for access tokens
-   `configTrust.trust[].port`: port, (keyStone) to exchange tokens for access tokens
-   `configTrust.trust[].id`: id of trust to use in update action rule
-   `configTrust.trust[].user`: provisioned user for trust in Keystone
-   `configTrust.trust[].password`: provisioned password for trust in Keystone
-   `configTrust.trust[].service`: provisioned service for trust in Keystone

Options for NGSI (for update action)

-   `config.castTypes`: cast or not attribute values to expected type conform NGSI-v2 (false by default)

URL format for mongoDB could be found at `http://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html`


Changelog
1.39.0 (October 14th, 2025)

- Add: allow use all phone number formats (local, short, international) with smpp rules (#861)
- Add: new defaultCountryCode configuration field (associated env var PERSEO_SMPP_DEFAULT_COUNTRY_CODE ) (#861)

1.38.0 (October 10th, 2025)

- Add: allow mix sms and smpp rules in the same instance of perseo
- Add: allow use multiple destinataries in action of SMPP rules
- Fix: SMPP fields for from and to
- Fix: return callback in action SMPP in all cases (to continue process further sms rules when error)
- Upgrade smpp dependency from 0.3.1 to 0.6.0-rc.4
- Upgrade: nodemailer dependency from from 6.9.16 to 7.0.7 (solving vulnerability)
- Set Nodejs 20 as minimum version in packages.json (effectively removing Nodev16 and Nodev18 from supported versions)

1.37.0 (September 24th, 2025)

- Add PERSEO_SMTP_TLS_CA, PERSEO_SMTP_TLS_KEY, PERSEO_SMTP_TLS_CERT, PERSEO_SMTP_TLS_MIN_VERSION and PERSEO_SMTP_TLS_SERVER_NAME env vars for global config (#841)
- Add: allow send email bodies as HTML (keeping also text) when template is expanded to html (#837)
- Upgrade NodeJS version from 16-slim to 24-bullseye-slim in Dockerfile

1.36.0 (April 7th, 2025)

- Add: max time detection for non signal rules to not see an entity as silent due to is death (#824)
- Add: reportIntervalAttr, maxTimeDetectionAttr and maxTimeDetection optional fields to non signal rules (#824)
- Add: env var PERSEO_NON_SIGNAL_MAX_TIME_DETECTION to set global nonSignalMaxTimeDetection (#824)
- Fix: remove cache token when action rule uses trust at update or remove rule (#831)
- Fix: prevent Perseo from crashing on startup with malformed configTrust file (#830)

1.35.0 (March 4th, 2025)

- Add: env var PERSEO_MONGO_DB_NAME to set mongo DB name (#819)
- Upgrade express from 4.19.2 to 4.21.2
- Upgrade body-parser dep from 1.18.2 to 1.20.3
- Upgrade nodemailer dep from 6.9.13 to 6.9.16
- Upgrade: nanoid dependency from 3.3.4 to 3.3.8 (solving vulnerability CVE-2024-55565)
- Fix: deprecated mongo warnings about save and remove methods (#806)
- Fix: remove unnecessary env vars (PERSEO_MONGO_HOST and PERSEO_CORE_URL) from Dockerfile (#810)

1.34.0 (July 30th, 2024)

- Fix: smtpConfig after error in emailAction rule (#798)

1.33.0 (July 29th, 2024)

- Fix: do not search in entities collection with strict mode by non_signal checker (#793)
- Fix: do not try cast entity id and entity type in update rules when expand parameters of rule to avoid BadRequest en CB tries to update that entities (#791)
- Fix: do not invoke calback twice when error about trust not found in trustConf (#790)

1.32.0 (June 13th, 2024)

- Add: Send multiple sms when multiple destination in sms action
- Fix: upgrade rule in mongo using service and subservice to avoid match with other rule (#783)

1.31.0 (June 11th, 2024)

- Fix: release transport resources of nodemailer after send email
- Upgrade express dep from 4.18.1 to 4.19.2 due to a vulnerability
- Upgrade nodemailer dep from 6.6.5 to 6.9.13 due to a vulnerability

1.30.0 (February 26th, 2024)

- Fix: do not stringify template in postAction if it is already a string (#763)

1.29.0 (December 1st, 2023)

- Add: watch changes and reload configTrust file (#757)

1.28.0 (October 4th, 2023)

- Add: new approach to handle trust auth (urbo-deployer#868)
- Add: allow access entities using NGSIv2 API for non_signal rules (new setting nonSignalByAPI / PERSEO_CHECK_NON_SIGNAL_BY_API) (#549)
- Add: allow use expandVar with JSON objects (#703)
- Add: apply expandVar with JSON.parse to all fields of all actions (sms, smpp, email, post, update) (#746)
- Fix: check domain before access domain
- Fix: expandVar return a 'null' instead of null (#746)
- Fix: expandVar for single string values with null, boolean, number (#746)
- Fix: smtp and smpp logs (#738)
- Fix: check timer string in lower case in TimedRule checker
- Fix: dep of chai get-func-name to 2.0.0 to avoid installation errors
- Remove: field "authentication" support in update action
- Remove: config.authentication (along with related env vars PERSEO_AUTHENTICATION_*)
- Remove: support for ngsv1 notifications (#714)
- Remove: ngsiv1 support for updateAction (#714)
- Remove: RPM stuff

1.27.0 (May 18th, 2023)

- Add: support to run tests with node 18
- Fix: do not flat in to map object or array attributes of event (#720)
- Remove: unnecessary event data transformations (stripped, location parse, datetime parse) (#706, #712)
- Upgrade uuid dep from ~1.4.2 to 9.0.0
- Set Nodejs 16 as minimum version in packages.json (effectively removing Nodev12 and Nodev14 from supported versions)

1.26.0 (October 27, 2022)

- Fix: update actions with filter of timed based rules are not tried to execute after first failure updating entities (#708)


1.25.0 (July 1st, 2022)

- Add: full detail about batch update error or success in logs
- Fix: check rule.name match with EPL rule.text ruleName when update a text rule (#699)
- Fix: ensure datetime attrs from entity for nosignal events are in ISO format instead of epoc
- Fix: expanding an object variable in action of rule (#692)
- Fix: for EPL with context + insert + select rule (without expression) (#652)
- Fix: format of internalCurrentTime fixed to ISO 8601
- Fix: typo in non signal event internalCurentTime -> internalCurrentTime
- Upgrade NodeJS version from 14-slim to 16-slim in Dockerfile

1.24.0 (May 26th, 2022)

- Fix: changed detection of rule loop by checking max number of rule executions triggered by the same notification (with the same correlator) (new env var PERSEO_MAX_RULES_BY_CORR) (#678)
- Fix: replace deprecated dep shortid by nanoid
- Upgrade ngsijs dep from ~1.3.0 to 1.4.1
- Upgrade express dep from ~4.16.4 to 4.18.1
- Upgrade uuid dep from ~1.4.2 to 3.4.0

1.23.0 (May 11th, 2022)

- Add: keep original array and object value in event sent to perseo-core (#662)
- Set Nodejs 12 as minimum version in packages.json (effectively removing Nodev10 from supported versions)    

config.js
/*
 * Copyright 2015 Telefonica Investigación y Desarrollo, S.A.U
 *
 * This file is part of perseo-fe
 *
 * perseo-fe is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * perseo-fe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with perseo-fe.
 * If not, see http://www.gnu.org/licenses/.
 *
 * For those usages not covered by the GNU Affero General Public License
 * please contact with::[contacto@tid.es]
 */

'use strict';

var config = {};

/**
 * Default log level. Can be one of: 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'
 * @type {string}
 */
config.logLevel = 'INFO';

/**
 * Configures the exposed API.
 */
config.endpoint = {
    host: 'localhost',
    port: 9090,
    rulesPath: '/rules',
    actionsPath: '/actions/do',
    noticesPath: '/notices',
    vrPath: '/m2m/vrules',
    checkPath: '/check',
    versionPath: '/version',
    logPath: '/admin/log',
    metricsPath: '/admin/metrics'
};

/**
 * Is Master this one?
 */
config.isMaster = true;

/**
 * Delay (in milliseconds) for slave to execute an action
 */
config.slaveDelay = 500;

/**
 * DB Configuration.
 */
config.mongo = {
    // The URI to use for the database connection. It supports replica set URIs.
    // mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
    // I.e.: 'mongodb://user:pass@host1:27017,host2:27018,host3:27019/cep?replicaSet=myrep'
    url: 'mongodb://localhost:27017/cep'
};

/**
 * OrionDB Configuration.
 */
config.orionDb = {
    url: 'mongodb://localhost:27017/orion',
    collection: 'entities',
    prefix: 'orion',
    batchSize: 500
};

/**
 * EPL core options
 *
 * interval is the time in milliseconds between refreshing rules
 * at core. Each <<interval>> ms, the rules are sent to core.
 */
config.perseoCore = {
    rulesURL: 'http://localhost:8080/perseo-core/rules',
    noticesURL: 'http://localhost:8080/perseo-core/events',
    interval: 60e3 * 5
};
/**
 * NEXT EPL core options (with HA)
 */
config.nextCore = {
    //
    // Note this parameter is empty, so Perseo will not use HA by default
    //
    /*
    rulesURL : 'http://next-core:8080/perseo-core/rules',
    noticesURL : 'http://next-core:8080/perseo-core/events'
    */
};

/**
 * SMTP endpoint options
 */
config.smtp = {
    port: 25,
    host: 'smtpserver',
    secure: false,
    /*
     auth: {
     user: 'abc',
     pass: 'xyz'
     }
     ,
    */
    tls: {
        // keep rejectUnauthorized to false when secure is also false
        rejectUnauthorized: false
    }
};

/**
 * SMS endpoint options
 */
config.sms = {
    URL: 'http://sms-endpoint/smsoutbound',
    API_KEY: '',
    API_SECRET: '',
    from: 'tel:22012;phone-context=+34'
};

/**
 * SMPP endpoint options
 */
config.smpp = {
    host: '',
    port: '',
    systemid: '',
    password: '',
    from: '346666666',
    enabled: false,
    defaultCountryCode: '34'
};

/**
 * Orion (Context Broker) endpoint options
 */
config.orion = {
    URL: 'http://orion-endpoint:1026'
};

/**
 * Pep (proxy for Context Broker) endpoint options
 */
config.pep = {
    URL: 'http://pep-endpoint:1026'
};

/**
 * Collections
 * @type {String}
 */
config.collections = {
    rules: 'rules',
    executions: 'executions'
};

/**
 * Executions TTL
 *
 * Number of seconds to expire a document in 'executions'
 * @type {Number}
 *
 */
config.executionsTTL = 1 * 24 * 60 * 60;

/**
 * Constants for missing header fields for service (Fiware-servicepath) DEFAULT_SUBSERVICE
 * and tenant (Fiware-service) DEFAULT_TENANT
 *
 * @type {String}
 */
config.DEFAULT_SUBSERVICE = '/';
config.DEFAULT_SERVICE = 'unknownt';

/**
 * CheckDB configuration
 *
 * delay Number of milliseconds to check DB connection
 * @type {Number}
 *
 * reportInterval Number of milliseconds to report a problem with DB connection in logs
 * @type {Number}
 */
config.checkDB = {
    delay: 5e3,
    reconnectTries: 1e3,
    reconnectInterval: 5e3,
    bufferMaxEntries: 5
};

/**
 * Rest base for changing access to twitter
 *
 * It should not be changed in normal conditions
 *
 * restBase:  URL
 * @type {String}
 *
 */
config.restBase = null;

/**
 * Cast attribute values in updateAction using NGSIv2 types
 */
config.castTypes = false;

/**
 * Maximum number of rule executions triggered by the same notification (with the same correlator) until refuse execute a rule
 */
config.maxRulesByCorr = 20;

/**
 * Check nonSignal rules using API ngsiv2 or just access to mongo
 * @type {Boolean}
 */
config.nonSignalByAPI = false;

/**
 * Max time detection for check nonSignal rules by mongo in seconds. By default 30 days = 2592000 seconds
 * @type {Number}
 */
config.nonSignalMaxTimeDetection = 2592000;

module.exports = config;
